### ISSUE TITLE
eopkg 4.0 test plan

### ISSUE THREAD

**Issue by livingsilver94** [getsolus/eopkg#22]
The transition to Python 3 is pretty invasive, and on top of it we also removed the piksemel dependency in favor of lxml. To find out bugs and prove we're production-ready, here's our User Acceptance Test suite:

> :warning: **That might break your Solus installation**. Unless you're experienced with [boot rescue](https://help.getsol.us/docs/user/troubleshooting/boot-rescue/) and the terminal, experiment in a virtual machine instead of in your local machine directly.

## Phase 1

We'll first create a fake system root in a directory of your choice. We'll call this directory `ROOT`.

```bash
sudo ./eopkg-cli ur -D ROOT # This will automatically add the Solus stable repository, and will also synchronize it.
sudo ./eopkg-cli lc -l -D ROOT # List all components (AKA groups) in the Solus repository.
sudo ./eopkg-cli la -l -D ROOT # List all available packages in the Solus repository.
sudo ./eopkg-cli it -c system.base -D ROOT # Install the base system.
# The installation above will fail because of two missing symlinks. Fix them.
sudo ln -s lib64 ROOT/lib
sudo ln -s ../run ROOT/var/run
sudo ./eopkg-cli it -c system.base -D ROOT # This time it will complete.
```

## Phase 2

Team members have created a package `eopkg4` that can be installed in parallel with the Python 2 eopkg (the `pisi` package). Note: software center is untested and may be broken if you install `eopkg4`. This may include the ability to install and update 3rd-party packages! So you must be familiar with the terminal to test `eopkg4`.

### Install eopkg4
```bash
# Add the Grocery repo
sudo eopkg ar Grocery https://repo-cdn.gzgz.dev/eopkg-index.xml.xz
# Install the Python 3 eopkg!
sudo eopkg it eopkg4
```

As of the current date (2023-10-26), some manual steps are needed to use `eopkg4`:
1. Enter the directory `/var/lib/eopkg/info/`. There should be a file called `files.db` in it.
2. Create a backup of the file by running `sudo mv files.db files.db.3`
3. Now, run `sudo eopkg4 rdb` to rebuild the files database.
4. The rebuild database step should finish successfully. Now, you're ready to use the Python 3 eopkg! Just execute `eopkg4` in place of `eopkg`.

Any thing `eopkg` can do, `eopkg4` should be able to do it too. If you notice any discrepancies, please open an issue or comment below.

### Switch Back/Between Python 2 eopkg
1. `sudo eopkg4 dc` to clear the cache.
2. Enter the directory `/var/lib/eopkg/info`. You should have a db backup `files.db.3`. Restore that backup by running `cp -f files.db.3 files.db`.
3. Run `sudo eopkg rdb` to rebuild the files database.
4. The rebuild database step should finish successfully. You're back to the Python 2 eopkg (and software center should also work) now! 

### Known Issues
1. `eopkg4 help` displays fewer commands than `eopkg help`, but all of the commands are still there. They just don't appear there for some reason.

If any issues arise in this process, feel free to comment below or open a new issue and ping @GZGavinZhao.

**Comment by GZGavinZhao** [comment]
Here's my report after about one week of usage:
1. `remove-repo`, `disable-repo`, `enable-repo` don't work.
2. `eopkg lu` doesn't list upgradable packages, but running `eopkg up` is works fine.

**Comment by GZGavinZhao** [comment]
`eopkg index` doesn't seem to work:
```
> sudo eopkg4 index --skip-signing . --debug
Building index of eopkg files under .

Adding package to index: llvm-devel-16.0.6-110-1-x86_64.eopkg                   
System error. Program terminated.
<class 'ValueError'>: Invalid attribute name 'xml:lang'
Please use 'eopkg help' for general help.

Traceback:
  File "/usr/bin/eopkg4", line 33, in <module>
    sys.exit(load_entry_point('eopkg==4.0', 'console_scripts', 'eopkg')())
  File "/usr/lib/python3.10/site-packages/pisi/scripts/eopkg.py", line 76, in main
    cli.run_command()
  File "/usr/lib/python3.10/site-packages/pisi/cli/pisicli.py", line 131, in run_command
    self.command.run()
  File "/usr/lib/python3.10/site-packages/pisi/cli/index.py", line 82, in run
    index(
  File "/usr/lib/python3.10/site-packages/pisi/api.py", line 808, in index
    index.write(output, sha1sum=True, compress=compression, sign=sign)
  File "/usr/lib/python3.10/site-packages/pisi/pxml/autoxml.py", line 524, in write
    self.encode(self.rootNode(), errs)
  File "/usr/lib/python3.10/site-packages/pisi/pxml/autoxml.py", line 365, in encode
    encode_member(self, node, errs)
  File "/usr/lib/python3.10/site-packages/pisi/pxml/autoxml.py", line 624, in encode
    encode_a(node, getattr(self, name, None), errs)
  File "/usr/lib/python3.10/site-packages/pisi/pxml/autoxml.py", line 858, in encode
    encode_item(listnode, item, errs)
  File "/usr/lib/python3.10/site-packages/pisi/pxml/autoxml.py", line 792, in encode
    obj.encode(classnode, errs)
  File "/usr/lib/python3.10/site-packages/pisi/pxml/autoxml.py", line 363, in encode
    base.encode(self, node, errs)
  File "/usr/lib/python3.10/site-packages/pisi/pxml/autoxml.py", line 365, in encode
    encode_member(self, node, errs)
  File "/usr/lib/python3.10/site-packages/pisi/pxml/autoxml.py", line 624, in encode
    encode_a(node, getattr(self, name, None), errs)
  File "/usr/lib/python3.10/site-packages/pisi/pxml/autoxml.py", line 916, in encode
    obj.encode(node, errs)
  File "/usr/lib/python3.10/site-packages/pisi/pxml/autoxml.py", line 91, in encode
    xmlext.setNodeAttribute(newnode, "xml:lang", key)
  File "/usr/lib/python3.10/site-packages/pisi/pxml/xmlext.py", line 44, in setNodeAttribute
    node.attrib[attrname] = value
  File "src/lxml/etree.pyx", line 2446, in lxml.etree._Attrib.__setitem__
  File "src/lxml/apihelpers.pxi", line 586, in lxml.etree._setAttributeValue
  File "src/lxml/apihelpers.pxi", line 1764, in lxml.etree._attributeValidOrRaise
```

**Comment by joebonrichie** [comment]
lxml is slow as balls unfortunately

Benchmarks (with 12 updates available):

`eopkg install libflac-dbginfo --dry-run --ignore-revdeps-of-deps`
Before: 0m1.116s | Now: 0m3.799s

`eopkg install libflac-dbginfo --dry-run
Before: 0m1.528s | Now: 0m5.900s

`eopkg lu`
Before: 0.455ms | Now: 0m1.344s

` eopkg upgrade --dry-run`
Before: 0m2.292s | Now: 0m4.442s

`update-repo --force` was run when switching between py2 and py3 versions to update the pickle cache protocol.

**Comment by ermo** [comment]
@livingsilver94 

This has been superseded by eopkg4-bin, hasn't it? What would you like to do with this issue...? Do we keep the associated draft PR open for development/testing purposes or...?

**Comment by ermo** [comment]
I'm closing this in favour of the other open issues.

This issue got us started and thank you to @livingsilver94 for that!
